docs/git-chat.1
.TH git-chat 1 "@CMAKE_COMPILATION_DATE@" "git-chat @CMAKE_PROJECT_VERSION_MAJOR@.@CMAKE_PROJECT_VERSION_MINOR@.@CMAKE_PROJECT_VERSION_PATCH@" "git-chat manual"

.SH NAME
git-chat \- use a Git repository as a chatroom


.SH SYNOPSIS
.sp
.nf
\fIgit-chat\fR [<options>] <command> [<args>]
\fIgit-chat\fR [\-h | \-\-help]
\fIgit-chat\fR [\-v | \-\-version]


.SH DESCRIPTION
\fIgit-chat\fR is a secure messaging application build on top of the \fIGit\fR suite. It leverages Git's ability to track content, using commit objects as messages and branches as message channels. \fIgit-chat\fR also leverages GnuPG for message signing and encryption.

If you're using git-chat for messaging, keep in mind that it should not be used as secure and tamper proof communication. It suffers from potential tampering (lost messages, erased history, etc.). Please use at your own risk.


.SH OPTIONS
.TP
\-\-passphrase-fd <n>
When a passphrase is required, read the passphrase from the given file descriptor. Only the first line will be read from the file descriptor.

This option is primarily used to override the GPG engine pinentry functionality when interactive passphrase entry isn't possible (e.g. when used in scripts).

.TP
\-\-passphrase-file <file>
When a passphrase is required, read the passphrase from the given file. Only the first line will be read from the file.

This option is primarily used to override the GPG engine pinentry functionality when interactive passphrase entry isn't possible (e.g. when used in scripts).

Storing passwords in plaintext is ill advised.

.TP
\-\-passphrase <pass>
When a passphrase is required, use the supplied passphrase.

This option is primarily used to override the GPG engine pinentry functionality when interactive passphrase entry isn't possible (e.g. when used in scripts).

This option should be used with care; passwords can be leaked accidentally if this option is misused.

.TP
\-h, \-\-help
Print a simple synopsis, along with any common subcommands and options.

.TP
\-v, \-\-version
Print the current version for git-chat and its dependencies.


.SH SUBCOMMANDS
.TP
\fBgit-chat-channel\fR(1)
Create and manage message channels.

.TP
\fBgit-chat-get\fR(1)
Fetch new messages and channels from remote repositories.

.TP
\fBgit-chat-init\fR(1)
Initialize a new git-chat messaging space.

.TP
\fBgit-chat-import-key\fR(1)
Import GnuPG public keys into a channel.

.TP
\fBgit-chat-message\fR(1)
Create new gpg-encrypted messages.

.TP
\fBgit-chat-publish\fR(1)
Publish new messages to remote repositories.

.TP
\fBgit-chat-read\fR(1)
Display and format messages in a channel.


.SH FILE/DIRECTORY LAYOUT
@DOCS_FILE_DIRECTORY_LAYOUT_SECTION@


.SH CONFIGURATION
The config file has a format similar to INI or TOML, but simplified. It consists of key-value pairs, organized in sections.

.EX
    [ section_a ]
        key_1 = value
        key_2 = valueb

    [ section_b ]
        key_1 = 123
        key_a = 321
.EE

.TP
.B Recognized Configuration Properties
Each channel may have it's own configuration. The configurations are organized by channel name in the following manner:

.EX
[channel.<rev>]
    name = <channel alias>
    createdby = <channel creator>
    description = <description>
.EE


.SH ENVIRONMENT VARIABLES
\fIgit-chat\fR uses a number of environment variables that allow you to configure how the application behaves.

.TP
\fBGIT_CHAT_LOG_LEVEL\fR
Display additional logging to the console. May have any of the following values, ordered decreasing in verbosity level:
.RS
.IP \[bu]
ALL
.IP \[bu]
TRACE
.IP \[bu]
DEBUG
.IP \[bu]
INFO
.IP \[bu]
WARN
.IP \[bu]
ERROR
.IP \[bu]
NONE
.RE

.TP
\fBGIT_CHAT_PAGER\fR, \fBGIT_PAGER\fR, \fBPAGER\fR
When output is being paged (\fIgit-chat-read\fR, for example), these environment variables may be used to specify an alternate paging program. The variable value must be the absolute path to the executable (e.g. /usr/bin/cat).


.SH EXTENDING GIT-CHAT
\fIgit-chat\fR follows a similar extension model to Git, where executables located on the PATH that are prefixed with \fBgit-chat-\fR will be invoked when \fBgit chat <extension name>\fR is run at the command line. This allows you to build custom plugins in your language of choice.


.SH SEE ALSO
\fBgit-chat-init\fR(1)


.SH REPORTING BUGS
@DOCS_REPORTING_BUGS_SECTION@


.SH AUTHOR
@DOCS_AUTHORS_SECTION@


src/git-chat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "builtin.h"
#include "run-command.h"
#include "fs-utils.h"
#include "gnupg/gpg-common.h"
#include "version.h"
#include "parse-options.h"
#include "utils.h"

static const struct usage_string main_cmd_usage[] = {
		USAGE("git chat [<options>] <command> [<args>]"),
		USAGE("git chat [-h | --help]"),
		USAGE("git chat (-v | --version)"),
		USAGE_END()
};

static struct cmd_builtin *get_builtin(const char *);
static int run_builtin(struct cmd_builtin *, int, char *[]);
static int run_extension(const char *, int, char *[]);
static char *find_extension(char *);
static void print_version(void);

int main(int argc, char *argv[])
{
	int gpg_pass_fd = -1;
	char *gpg_pass_file = NULL;
	char *gpg_pass = NULL;

	int show_help = 0;
	int show_version = 0;

	const struct command_option main_cmd_options[] = {
			OPT_GROUP("commands"),
			OPT_CMD("channel", "create and manage communication channels", NULL),
			OPT_CMD("message", "create messages", NULL),
			OPT_CMD("publish", "publish messages to the remote server", NULL),
			OPT_CMD("get", "download messages", NULL),
			OPT_CMD("read", "display, format and read messages", NULL),
			OPT_CMD("import-key", "import a GPG key into the current channel", NULL),
			OPT_CMD("config", "configure a channel", NULL),

			OPT_GROUP("options"),
			OPT_LONG_INT("passphrase-fd", "read passphrase from file descriptor", &gpg_pass_fd),
			OPT_LONG_STRING("passphrase-file", "file", "read passphrase from file", &gpg_pass_file),
			OPT_LONG_STRING("passphrase", "pass", "use string as passphrase", &gpg_pass),
			OPT_BOOL('h', "help", "show usage and exit", &show_help),
			OPT_BOOL('v', "version", "output version information and exit", &show_version),
			OPT_END()
	};

	// show usage and return if no arguments provided
	if (argc < 2) {
		show_usage_with_options(main_cmd_usage, main_cmd_options, 0, NULL);
		return 0;
	}

	argc = parse_options(argc, argv, main_cmd_options, 1, 1);
	if (show_help) {
		show_usage_with_options(main_cmd_usage, main_cmd_options, 0, NULL);
		return 0;
	}

	if (show_version) {
		print_version();
		return 0;
	}

	if (gpg_pass_fd >= 0 && (gpg_pass_file || gpg_pass)) {
		show_usage_with_options(main_cmd_usage, main_cmd_options, 1,
				"error: --passphrase-fd cannot be combined with --passphrase-file or --passphrase");
		return 1;
	}
	if (gpg_pass_file && (gpg_pass_fd >= 0 || gpg_pass)) {
		show_usage_with_options(main_cmd_usage, main_cmd_options, 1,
				"error: --passphrase-file cannot be combined with --passphrase-fd or --passphrase");
		return 1;
	}
	if (gpg_pass && (gpg_pass_file || gpg_pass_fd >= 0)) {
		show_usage_with_options(main_cmd_usage, main_cmd_options, 1,
				"error: --passphrase cannot be combined with --passphrase-fd or --passphrase-file");
		return 1;
	}

	// configure gpgme passphrase loopback
	if (gpg_pass_fd >= 0)
		gpgme_configure_passphrase_loopback(gpgme_pass_fd_cb, &gpg_pass_fd);
	if (gpg_pass_file)
		gpgme_configure_passphrase_loopback(gpgme_pass_file_cb, gpg_pass_file);
	if (gpg_pass)
		gpgme_configure_passphrase_loopback(gpgme_pass_cb, gpg_pass);

	//delegate commands
	struct cmd_builtin *builtin = get_builtin(argv[0]);
	if (builtin == NULL) {
		//check if any extensions exist
		char *extension_on_path = find_extension(argv[0]);
		if (extension_on_path) {
			LOG_INFO("builtin: extension '%s' found on the PATH", extension_on_path);

			int status = run_extension(extension_on_path, argc - 1, argv + 1);
			free(extension_on_path);
			return status;
		}

		show_usage_with_options(main_cmd_usage, main_cmd_options, 1, "error: unknown command or option '%s'", argv[0]);
		return 1;
	}

	return run_builtin(builtin, argc, argv);
}

static struct cmd_builtin *get_builtin(const char *s)
{
	struct cmd_builtin *builtin = registered_builtins;
	while (builtin->cmd != NULL) {
		if (!strcmp(s, builtin->cmd))
			return builtin;

		builtin++;
	}

	return NULL;
}

static int run_builtin(struct cmd_builtin *builtin, int argc, char *argv[])
{
	init_gpgme_openpgp_engine();

	LOG_INFO("builtin: executing %s builtin", builtin->cmd);
	return builtin->fn(argc, argv);
}

static int run_extension(const char *s, int argc, char *argv[])
{
	LOG_INFO("builtin: executing %s as an extension", s);

	struct child_process_def cmd;
	child_process_def_init(&cmd);
	cmd.executable = s;
	for (int i = 0; i < argc; i++)
		argv_array_push(&cmd.args, argv[i], NULL);

	int status = run_command(&cmd);
	child_process_def_release(&cmd);

	return status;
}

static char *find_extension(char *extension_name)
{
	struct strbuf filename;
	strbuf_init(&filename);

	strbuf_attach_str(&filename, "git-chat-");
	strbuf_attach_str(&filename, extension_name);
	char *path = find_in_path(filename.buff);
	strbuf_release(&filename);

	if (!path) {
		return NULL;
	}

	if (!is_executable(path)) {
		return NULL;
	}

	return path;
}

static void print_version(void)
{
	fprintf(stdout, "git-chat version %s\n", GIT_CHAT_VERSION);

	struct child_process_def cmd;
	child_process_def_init(&cmd);

	cmd.git_cmd = 1;
	argv_array_push(&cmd.args, "--version", NULL);
	run_command(&cmd);
	child_process_def_release(&cmd);

	fprintf(stdout, "gpgme library version %s\n",
			get_gpgme_library_version());
}
